lib/portlayer/exec/container.go

```diff
@@ -591,40 +602,57 @@ func (c *Container) Remove(ctx context.Context, sess *session.Session) error {
 	// enable Destroy
 	c.vm.EnableDestroy(ctx)
 
-	//removes the vm from vsphere, but detaches the disks first
+	concurrent := false
+	// if DeleteExceptDisks succeeds on VC, it leaves the VM orphan so we need to call Unregister
+	// if DeleteExceptDisks succeeds on ESXi, no further action needed
+	// if DeleteExceptDisks fails, we should call Unregister and only return an error if that fails too
+	//		Unregister sometimes can fail with ManagedObjectNotFound so we ignore it
 	_, err = c.vm.WaitForResult(ctx, func(ctx context.Context) (tasks.Task, error) {
 		return c.vm.DeleteExceptDisks(ctx)
 	})
 	if err != nil {
 		f, ok := err.(types.HasFault)
 		if !ok {
+			log.Warnf("DeleteExceptDisks failed with non-fault error %s for %s.", err, c)
+
 			c.updateState(existingState)
 			return err
 		}
+
 		switch f.Fault().(type) {
 		case *types.InvalidState:
-			log.Warnf("container VM is in invalid state, unregistering")
+			log.Warnf("container VM %s is in invalid state, unregistering", c)
 			if err := c.vm.Unregister(ctx); err != nil {
-				log.Errorf("Error while attempting to unregister container VM: %s", err)
+				log.Errorf("Error while attempting to unregister container VM %s: %s", c, err)
 				return err
 			}
 		case *types.ConcurrentAccess:
 			// We are getting ConcurrentAccess errors from DeleteExceptDisks - even though we don't set ChangeVersion in that path
 			// We are ignoring the error because in reality the operation finishes successfully.
-			log.Warnf("DeleteExceptDisks failed with ConcurrentAccess error. Ignoring it.")
+			log.Warnf("DeleteExceptDisks failed with ConcurrentAccess error for %s. Ignoring it.", c)
+			concurrent = true
 		default:
-			log.Debugf("Fault while attempting to destroy vm: %#v", f.Fault())
+			log.Debugf("Unhandled fault while attempting to destroy vm %s: %#v", c, f.Fault())
+
 			c.updateState(existingState)
 			return err
 		}
 	}
 
+	if concurrent && c.vm.IsVC() {
```

Comment:
move this into vm.DeleteExceptDisks instead of inlined here

Reply:
Hmm why? The above 15 lines or so already doing similar thing (aka handling the err returned by DeleteExcept), moving this there will just hide this but won't change anything

Reply:
I slept on it and decided to move the whole corner case handling into DeleteExceptDisks :) Will refresh the PR

Reply:
And nope, not going to happen unless we change all the callers. Everyone wraped the DeleteExceptDisk with task package (so that fn. returns the task not the result). Since ConcurrentAccessError could be thrown by either Reconfigure or Destroy we also need to keep that info somewhere so that next retry could do the right thing. So unless I'm missing something (monday/haven't finished the first cup of coffee) I'll keep this PR intact